File: java/src/main/java/com/experian/dvs/client/phone/Configuration.java
package com.experian.dvs.client.phone;

import com.experian.dvs.client.common.Country;

import java.util.ArrayList;
import java.util.List;

/**
 * Configuration class for setting up the phone client.
 * Provides options for customizing phone-related API requests.
 */
public class Configuration extends com.experian.dvs.client.common.Configuration {

    public static final int DEFAULT_CACHE_VALUE_DAYS = 7;
    private final boolean metadata;
    private final String outputFormat;
    private final int cacheValueDays;
    private final boolean getPortedDate;
    private final boolean getDisposableNumber;
    private final List<Country> liveStatusForMobile;
    private final List<Country> liveStatusForLandline;

    /**
     * Initializes a new instance of the {@link Configuration} class using the specified builder.
     *
     * @param builder The builder containing the configuration settings.
     */
    protected Configuration(Builder builder) {
        super(builder);
        this.metadata = builder.metadata;
        this.outputFormat = builder.outputFormat;
        this.cacheValueDays = builder.cacheValueDays;
        this.getPortedDate = builder.getPortedDate;
        this.getDisposableNumber = builder.getDisposableNumber;
        this.liveStatusForMobile = builder.liveStatusForMobile;
        this.liveStatusForLandline = builder.liveStatusForLandline;
    }

    /**
     * Creates a new instance of the {@link Builder} class to configure the phone client.
     *
     * @param token The authentication token for the API.
     * @return A new instance of the {@link Builder} class.
     */
    public static Configuration.Builder newBuilder(String token) {
        return new Builder(token);
    }

    /**
     * Retrieves the output format setting for the configuration.
     *
     * @return The output format as a string.
     */
    public String getOutputFormat() {
        return outputFormat;
    }

    /**
     * Retrieves the cache value days setting for the configuration.
     *
     * @return The number of days to cache values.
     */
    public int getCacheValueDays() {
        return cacheValueDays;
    }

    /**
     * Retrieves the ported date inclusion setting for the configuration.
     *
     * @return True if the ported date is included; otherwise, false.
     */
    public boolean getGetPortedDate() {
        return getPortedDate;
    }

    /**
     * Retrieves the disposable number inclusion setting for the configuration.
     *
     * @return True if disposable numbers are included; otherwise, false.
     */
    public boolean getGetDisposableNumber() {
        return getDisposableNumber;
    }

    /**
     * Retrieves the list of countries for which live status is enabled for landline numbers.
     *
     * @return A list of {@link Country} objects.
     */
    public List<Country> getLiveStatusForLandline() {
        return liveStatusForLandline;
    }

    /**
     * Retrieves the metadata inclusion setting for the configuration.
     *
     * @return True if metadata is included; otherwise, false.
     */
    public boolean getMetadata() {
        return metadata;
    }

    /**
     * Retrieves the list of countries for which live status is enabled for mobile numbers.
     *
     * @return A list of {@link Country} objects.
     */
    public List<Country> getLiveStatusForMobile() {
        return liveStatusForMobile;
    }

    /**
     * Builder class for constructing a {@link Configuration} object with custom settings.
     */
    public static class Builder extends com.experian.dvs.client.common.Configuration.Builder {
        private boolean metadata;
        private String outputFormat = "";
        private int cacheValueDays = DEFAULT_CACHE_VALUE_DAYS;
        private boolean getPortedDate;
        private boolean getDisposableNumber;
        private List<Country> liveStatusForMobile = new ArrayList<Country>();
        private List<Country> liveStatusForLandline = new ArrayList<Country>();

        /**
         * Initializes a new instance of the {@link Builder} class with the specified token.
         *
         * @param token The authentication token for the API.
         */
        public Builder(String token) {
            super(token);
        }

        /**
         * Sets a custom transaction ID for API requests.
         *
         * @param transactionId The transaction ID to use.
         * @return The current {@link Builder} instance for method chaining.
         */
        public Builder setTransactionId(String transactionId) {
            super.setTransactionId(transactionId);
            return this;
        }

        /**
         * Includes metadata in the API response.
         *
         * @return The current {@link Builder} instance for method chaining.
         */
        public Builder includeMetadata() {
            this.metadata = true;
            return this;
        }

        /**
         * Sets the output format for the API response.
         *
         * @param outputFormat The output format to use.
         * @return The current {@link Builder} instance for method chaining.
         */
        public Builder useOutputFormat(String outputFormat) {
            this.outputFormat = outputFormat;
            return this;
        }

        /**
         * Sets the number of days to cache values.
         *
         * @param cacheValueDays The number of days to cache values.
         * @return The current {@link Builder} instance for method chaining.
         */
        public Builder useCacheValueDays(Integer cacheValueDays) {
            this.cacheValueDays = cacheValueDays;
            return this;
        }

        /**
         * Includes the ported date in the API response.
         *
         * @return The current {@link Builder} instance for method chaining.
         */
        public Builder includePortedDate() {
            this.getPortedDate = true;
            return this;
        }

        /**
         * Includes disposable numbers in the API response.
         *
         * @return The current {@link Builder} instance for method chaining.
         */
        public Builder includeDisposableNumber() {
            this.getDisposableNumber = true;
            return this;
        }

        /**
         * Sets the list of countries for which live status is enabled for mobile numbers.
         *
         * @param liveStatusForMobile The list of countries to enable live status for mobile numbers.
         * @return The current {@link Builder} instance for method chaining.
         */
        public Builder useLiveStatusForMobile(List<Country> liveStatusForMobile) {
            this.liveStatusForMobile = liveStatusForMobile;
            return this;
        }

        /**
         * Sets the list of countries for which live status is enabled for landline numbers.
         *
         * @param liveStatusForLandline The list of countries to enable live status for landline numbers.
         * @return The current {@link Builder} instance for method chaining.
         */
        public Builder useLiveStatusForLandline(List<Country> liveStatusForLandline) {
            this.liveStatusForLandline = liveStatusForLandline;
            return this;
        }

        /**
         * Builds the {@link Configuration} object with the specified settings.
         *
         * @return A new instance of the {@link Configuration} class.
         */
        @Override
        public Configuration build() {
            return new Configuration(this);
        }
    }
}

File: csharp/DVSClient/DVSClient/phone/Configuration.cs
using DVSClient.Common;

namespace DVSClient.Phone
{
    /// <summary>
    /// Configuration class for setting up the email validation client.
    /// </summary>
    public class Configuration : Common.Configuration
    {
        public const int DefaultCacheValueDays = 7;
        internal bool? Metadata { get; }
        internal string OutputFormat { get; }
        internal int? CacheValueDays { get; }
        internal bool? IncludeDisposableNumber { get; }
        internal bool? IncludePortedDate { get; }
        internal IEnumerable<Country> LiveStatusForMobile { get; }
        internal IEnumerable<Country> LiveStatusForLandline { get; }

        private Configuration(Builder builder) : base(builder)
        {
            Metadata = builder.Metadata;
            OutputFormat = builder.OutputFormat;
            CacheValueDays = builder.CacheValueDays;
            IncludePortedDate = builder.GetPortedDate;
            IncludeDisposableNumber = builder.GetDisposableNumber;
            LiveStatusForMobile = builder.LiveStatusForMobile;
            LiveStatusForLandline = builder.LiveStatusForLandline;
        }

        /// <summary>
        /// Creates a new builder instance for configuring the phone validation client.
        /// </summary>
        /// <param name="token">The API token to authenticate requests.</param>
        /// <returns>A new instance of the Builder class.</returns>
        public static Builder NewBuilder(string? token)
        {
            return new Builder(token);
        }

        public new class Builder : Common.Configuration.Builder
        {
            internal bool Metadata { get; private set; }
            internal string OutputFormat { get; private set; } = string.Empty;
            internal int CacheValueDays { get; private set; } = DefaultCacheValueDays;
            internal bool GetPortedDate { get; private set; }
            internal bool GetDisposableNumber { get; private set; }
            internal IEnumerable<Country> LiveStatusForMobile { get; private set; } = new List<Country>();
            internal IEnumerable<Country> LiveStatusForLandline { get; private set; } = new List<Country>();

            /// <summary>
            /// Initializes a new instance of the Builder class.
            /// </summary>
            /// <param name="token">The API token to authenticate requests.</param>
            public Builder(string? token) : base(token)
            {
            }

            /// <summary>
            /// Sets the maximum delay between retries for API requests.
            /// </summary>
            /// <param name="maxDelay">The maximum delay in milliseconds.</param>
            /// <returns>The current Builder instance for method chaining.</returns>
            public new Builder SetMaxDelay(int maxDelay)
            {
                base.SetMaxDelay(maxDelay);
                return this;
            }

            /// <summary>
            /// Sets the initial delay before retrying API requests.
            /// </summary>
            /// <param name="initialDelay">The initial delay in milliseconds.</param>
            /// <returns>The current Builder instance for method chaining.</returns>
            public new Builder SetInitialDelay(int initialDelay)
            {
                base.SetInitialDelay(initialDelay);
                return this;
            }

            /// <summary>
            /// Sets the maximum number of retries for API requests.
            /// </summary>
            /// <param name="maxRetries">The maximum number of retries.</param>
            /// <returns>The current Builder instance for method chaining.</returns>
            public new Builder SetMaxRetries(int maxRetries)
            {
                base.SetMaxRetries(maxRetries);
                return this;
            }

            /// <summary>
            /// Sets the timeout for API requests in seconds.
            /// </summary>
            /// <param name="timeouts">The timeout duration in seconds.</param>
            /// <returns>The current Builder instance for method chaining.</returns>
            public new Builder SetApiRequestTimeoutInSeconds(int timeouts)
            {
                base.SetApiRequestTimeoutInSeconds(timeouts);
                return this;
            }

            /// <summary>
            /// Sets the timeout for the HTTP client in seconds.
            /// </summary>
            /// <param name="timeouts">The timeout duration in seconds.</param>
            /// <returns>The current Builder instance for method chaining.</returns>
            public new Builder SetHttpClientTimeoutInSeconds(int timeouts)
            {
                base.SetHttpClientTimeoutInSeconds(timeouts);
                return this;
            }

            /// <summary>
            /// Sets a custom transaction ID for API requests.
            /// </summary>
            /// <param name="transactionId">The transaction ID to use.</param>
            /// <returns>The current Builder instance for method chaining.</returns>
            public new Builder SetTransactionId(string transactionId)
            {
                base.SetTransactionId(transactionId);
                return this;
            }

            /// <summary>
            /// Includes metadata in the API response.
            /// </summary>
            /// <returns>The current Builder instance for method chaining.</returns>
            public Builder IncludeMetadata()
            {
                Metadata = true;
                return this;
            }

            /// <summary>
            /// Specifies the phone number output format for the API response.
            /// </summary>
            /// <param name="outputFormat">The desired output format (e.g., NATIONAL).</param>
            /// <returns>The current Builder instance for method chaining.</returns>
            public Builder UseOutputFormat(string outputFormat)
            {
                OutputFormat = outputFormat;
                return this;
            }

            /// <summary>
            /// Sets the maximum ages of the cache for the phone number validation result.
            /// </summary>
            /// <param name="cacheValueDays">The maximum age of the cache.</param>
            /// <returns>The current Builder instance for method chaining.</returns>
            public Builder UseCacheValueDays(int cacheValueDays)
            {
                CacheValueDays = cacheValueDays;
                return this;
            }

            /// <summary>
            /// Includes the ported date of the phone number in the API response.
            /// </summary>
            /// <returns>The current Builder instance for method chaining.</returns>
            public Builder IncludePortedDate()
            {
                GetPortedDate = true;
                return this;
            }

            /// <summary>
            /// Includes information about whether the phone number is disposable in the API response.
            /// </summary>
            /// <returns>The current Builder instance for method chaining.</returns>
            public Builder IncludeDisposableNumber()
            {
                GetDisposableNumber = true;
                return this;
            }

            /// <summary>
            /// Specifies the countries for which live status should be checked for mobile numbers.
            /// </summary>
            /// <param name="liveStatusForMobile">A list of countries to check live status for mobile numbers.</param>
            /// <returns>The current Builder instance for method chaining.</returns>
            public Builder UseLiveStatusForMobile(IEnumerable<Country> liveStatusForMobile)
            {
                LiveStatusForMobile = liveStatusForMobile;
                return this;
            }

            /// <summary>
            /// Specifies the countries for which live status should be checked for landline numbers.
            /// </summary>
            /// <param name="liveStatusForLandline">A list of countries to check live status for landline numbers.</param>
            /// <returns>The current Builder instance for method chaining.</returns>
            public Builder UseLiveStatusForLandline(IEnumerable<Country> liveStatusForLandline)
            {
                LiveStatusForLandline = liveStatusForLandline;
                return this;
            }

            /// <summary>
            /// Builds the Configuration object with the specified settings.
            /// </summary>
            /// <returns>A Configuration object with the configured settings.</returns>
            public override Configuration Build()
            {
                return new Configuration(this);
            }
        }
    }
}
